Create working cmake files for example files

danglevm opened the pull request:
# Purpose
- Add working cmakelists for example files

- Notion task: https://www.notion.so/uworbital/6375c5c5dcf1485e8a6c47f18d419ae5?v=72baf90d01f643b09d77418bd3f53b14&p=062e93d0a1b54c608dcdf98fa46a720c&pm=s

# New Changes
- Added `Examples` build option in top-level directory CMakeLists
- Added options in obc/CMakeLists to build specified example in obc/examples
- Added instructions for how to build an example in README
- Make `-Werror` only apply for OBC build (test that it still errors on warnings by adding unused variables)
- Move and separate tests under comms_link_mgr to examples 
- Delete unused makefiles in examples 

# Testing
- Commented out initLogger() function from `test_app_rtc/main.c`
- Manually build all the example files from `obc/examples` with new CMakeLists.

# Outstanding Changes
- Make `-Werror` apply to all example builds once all example builds are fixed
- Make obc github workflows independent for each board
- Fix errors from example source files. Listed here:
- `test_app_fram_persist`, `test_app_rs` has unresolved warnings.
- `test_app_freertos_posix` has unresolved warnings and source file errors.
 - `test_app_mpu6050`, `test_app_cc1120_spi` has source file errors.
 - `test_app_rtc` - location of initLogger is changed so need update from source file.
 -  `test_app_fram_spi`, `test_app_reliance_edge_sd`, `test_app_uart_rx`, `test_app_uart_tx`, `dma_spi_demo` has no errors.

danglevm committed:
create first example of CMake for dma spi

danglevm committed:
non-working version to build demo SPI

danglevm committed:
add more include directories and source files

danglevm committed:
non-working version

danglevm committed:
remove sources and add absolute path

danglevm committed:
switch to add library and add files to compile

danglevm committed:
add more source files and include more directories

danglevm committed:
cleanup

danglevm committed:
add example build options

danglevm committed:
modify obc and root cmakelists

- add Examples build_type to root cmakelists
- add header directories for MPU6050 and POSIX compilation

danglevm committed:
Add Example files CMake build guides

danglevm committed:
Add bullet points

dgobalak commented on obc/examples/dma_spi_demo/CMakeLists.txt:
Is this change supposed to be here?

danglevm replied:
Oh no, it shouldn't be. I'll have it deleted

dgobalak commented on obc/CMakeLists.txt:
```diff
@@ -15,7 +15,40 @@ else()
-add_executable(OBC-firmware.out main.c)
+if(${CMAKE_BUILD_TYPE} MATCHES OBC)
+	add_executable(OBC-firmware.out main.c)
+elseif(${CMAKE_BUILD_TYPE} MATCHES Examples)
+	if (${EXAMPLE_TYPE} MATCHES DMA_SPI)
+		add_executable(OBC-firmware.out examples/dma_spi_demo/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES FRAM_PERSIST)
+		add_executable(OBC-firmware.out examples/test_app_fram_persist/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES FRAM_SPI)
+		add_executable(OBC-firmware.out examples/test_app_fram_spi/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES FREERTOS_POSIX)
+		add_executable(OBC-firmware.out examples/test_app_freertos_posix/main.c)
+		target_include_directories(OBC-firmware.out PRIVATE examples/test_app_freertos_posix/include ${CMAKE_SOURCE_DIR}/libs/posix_freertos/include)
+		target_compile_definitions(OBC-firmware.out PRIVATE projCOVERAGE_TEST=0)
+	elseif(${EXAMPLE_TYPE} MATCHES LM75BD)
+		add_executable(OBC-firmware.out examples/test_app_lm75bd/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES MPU6050)
+		add_executable(OBC-firmware.out examples/test_app_mpu6050/main.c examples/test_app_mpu6050/source/mpu6050.c)
+		target_include_directories (OBC-firmware.out PRIVATE examples/test_app_mpu6050/include)
+	elseif(${EXAMPLE_TYPE} MATCHES RE_SD)
+		add_executable(OBC-firmware.out examples/test_app_reliance_edge_sd/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES RTC)
+		add_executable(OBC-firmware.out examples/test_app_rtc/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES UART_RX)
+		add_executable(OBC-firmware.out examples/test_app_uart_rx/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES UART_TX)
+		add_executable(OBC-firmware.out examples/test_app_uart_tx/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES VN100)
+		add_executable(OBC-firmware.out examples/vn100_demo/test_binary_reading/main.c)
+	else ()
+		message (FATAL_ERROR "Invalid example type: ${EXAMPLE_TYPE} ")
+	endif()
```
Can you make sure these get built in our GitHub workflow? That'd ensure the builds actually work. https://github.com/UWOrbital/OBC-firmware/blob/main/.github/workflows/obc_build.yml

dgobalak commented:
I'd just remove the posix freertos example from this tbh. It's not worth building imo since we don't use it

danglevm committed:
remove dma_spi cmakelist

danglevm committed:
add example build in github workflow; remove posix

danglevm committed:
move obc examples build workflow to new file

Yarik-Popov commented on README.md:
I would add instructions on how to edit the cmake and workflow files

danglevm replied:
Alright I'll do so then

Yarik-Popov commented on .github/workflows/obc_examples.yml:
The way this is set up can become very annoying and error-prone quickly. I would suggest looking at the GitHub actions dynamic matrix.

danglevm replied:
Got it, will do so.

Yarik-Popov committed:
Merge branch 'main' into dangle/create-working-cmake-files-for-example-files

Yarik-Popov committed:
Fix obc build

Yarik-Popov committed:
Add instructions on how to add an example to cmake, fix DMA_SPI example, make -Werror only apply to OBC

Yarik-Popov committed:
Add workflow for examples

Yarik-Popov committed:
Update example workflow

Yarik-Popov committed:
Update example workflow 2

Yarik-Popov committed:
Update example workflow 3

Yarik-Popov committed:
Update example workflow 4

Yarik-Popov committed:
Update example workflow 5

Yarik-Popov committed:
Add instructions on how to add to workflow, clean up README examples

Yarik-Popov committed:
Move comms link mgr tests to cc1120 spi and rs test examples

Yarik-Popov committed:
Remove example makefiles

Yarik-Popov committed:
Revert download toolchain

AmeenDurani committed:
HIL - LogSink (#295)

# Purpose

- Generate initial files for hil testing and add log sink functionality

# New Changes

- Implemented log sink class
- Added C++ functionality
- Created a python wrapper to use Pytest framework, and included examples
folder (**)

# Testing

![20240612_13h10m23s_grim](https://github.com/UWOrbital/OBC-firmware/assets/140562992/cc156a76-fa22-4dac-8f6c-0348b6d96de9)

Verified logs are getting saved to the specified file, gtest
framework can run the unit tests.

Navtajh04 committed:
[BUGFIX] Fix bootloop due to GNC mgr watchdog biting (#300)

Navtajh04 committed:
Add Timestamping to Logs (#299)

# Purpose
Explain the purpose of the PR here, including references to any existing
Notion tasks.
Add timestamps to logs so that when looking at the logs a while after
they have been reported you can see when they were reported
# New Changes
- Explain new changes

# Testing
- Explain tests that you ran to verify code functionality.
- Any functions that can be unit-tested should include a unit test in
the PR. Otherwise, explain why it cannot be unit-tested.
- Regression tested on launchpad, need someone on campus to test on rev1
or some board with RTC
# Outstanding Changes
- If there are non-critical changes (i.e. additional features) that can
be made to this feature in the future, indicate them here.
- Current precision is 1s which is fine since our telem is only
collected every min anyways. Once (if?) we have high freq logging we can
look into ms precision if needed
- GS tool to convert Unix time to datetime for logs received

kepler452b123 commented on obc/CMakeLists.txt:
```diff
@@ -28,7 +28,46 @@ endif()
+if(${CMAKE_BUILD_TYPE} MATCHES OBC)
+  add_compile_options(-Werror) # TODO: Move this back to download_arm_toolchain once all examples warnings are fixed
+	add_executable(OBC-firmware.out app/app_main.c)
+elseif(${CMAKE_BUILD_TYPE} MATCHES Examples)
+	if (${EXAMPLE_TYPE} MATCHES DMA_SPI)
+		add_executable(OBC-firmware.out examples/dma_spi_demo/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES FRAM_PERSIST)
+		add_executable(OBC-firmware.out examples/test_app_fram_persist/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES FRAM_SPI)
+		add_executable(OBC-firmware.out examples/test_app_fram_spi/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES LM75BD)
+		add_executable(OBC-firmware.out examples/test_app_lm75bd/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES MPU6050)
+		add_executable(OBC-firmware.out examples/test_app_mpu6050/main.c examples/test_app_mpu6050/source/mpu6050.c)
+		target_include_directories (OBC-firmware.out PRIVATE examples/test_app_mpu6050/include)
+	elseif(${EXAMPLE_TYPE} MATCHES RE_SD)
+		add_executable(OBC-firmware.out examples/test_app_reliance_edge_sd/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES RTC)
+		add_executable(OBC-firmware.out examples/test_app_rtc/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES UART_RX)
+		add_executable(OBC-firmware.out examples/test_app_uart_rx/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES UART_TX)
+		add_executable(OBC-firmware.out examples/test_app_uart_tx/main.c)
+	elseif(${EXAMPLE_TYPE} MATCHES VN100)
+		add_executable(OBC-firmware.out examples/vn100_demo/test_binary_reading/main.c)
+  elseif(${EXAMPLE_TYPE} MATCHES CC1120_SPI)
+    add_executable(OBC-firmware.out examples/test_app_cc1120_spi/main.c examples/test_app_cc1120_spi/cc1120_spi_tests.c)
+    target_include_directories (OBC-firmware.out PRIVATE examples/test_app_cc1120_spi)
+  elseif(${EXAMPLE_TYPE} MATCHES RS)
+    add_executable(OBC-firmware.out examples/test_app_rs/main.c)
```
Is it possible to add an option to build all the out files in case someone wants to test multiple things (e.g. for testing this PR)

kepler452b123 replied:
This would also mean you have to update the names from OBC-firmware.out to the specific test name. I think this would be more clear as well, and prevent overwriting other out files every time. e.g. add_executable(OBC-firmware-${EXAMPLE_TYPE}.out)

Yarik-Popov replied:
1. I would say it is more of a nice to have and I don't see it being used often. The workflow already builds all examples and this PR doesn't make sure that the examples actually work (check todo section). It might be possible to create an example to build all but I would say its not within the scope of the PR. This PR is just to setup individual building of examples so we don't have to copy and paste the main file from examples when test it thus speeding up testing times. 

2. I'll change it to display a different executable name

kepler452b123 commented on CMakeLists.txt:
is it necessary to update the top level cmake file? examples is under the obc directory so can this logic be added there to follow the logical order?

Yarik-Popov replied:
I disagree that it would be more logical to move the examples yop level cmakelist logic down to obc/cmakelist. The way this is done right now is more consistent as in the top level cmake we set the tool chain and languages and lower down we add main and other configuration. Otherwise, we would have different files setting different types of configuration only dependent on the type. 

Maybe later on we can restructure the cmake build as there is a lot of duplicate checks for the different types but not in the scope of this PR

kepler452b123 reviewed (commented):
some quick suggestions/things to think about. will test later

Commit:
added OUT_FILE_NAME to cmake file

Commit:
fixed cmake build errors

Commit:
fixed CMakeList.txt files in subdirs

Commit:
removed bloat files

Commit:
removed final bloat

Commit:
fixed syntax in CMakeLists.txt

Commit:
cmake change for .bin

Yarik-Popov commented on README.md:
````diff
@@ -192,6 +192,39 @@ cmake --build .
+- Add the following to the `OBC/CMakeLists.txt` above the comment that says `# ADD MORE EXAMPLES ABOVE THIS COMMENT`
+```
+elseif(${CMAKE_BUILD_EXAMPLE} MATCHES EXAMPLE_ID)
+	add_executable(OBC-firmware.out path_to_main_file_in_example)
````
This line needs to be changed to use the ${OUT_FILE_NAME} instead

Commit:
Updated Readme for examples

Commit:
Merge branch 'dangle/create-working-cmake-files-for-example-files'

Commit:
Fixing build

dgobalak commented on .github/workflows/hil_build.yml:
doesn't this file exist on main already?

dgobalak replied:
maybe try a `git pull` or a rebase?

Yarik-Popov replied:
Honestly, i tried to both rebase and pull, it was giving a bunch of merge conflicts around 20 each that need to manually be resolved so its better to create fresh PR at this point

Yarik-Popov committed:
Resolve merge conflict in obc examples

Yarik-Popov commented:
Being handled by #302
